WordPress/gutenberg
# Persist focus when tabbing back to the block toolbar

Fixes #24469

This PR adds a new experimental feature to the block toolbar so it "remembers" the last focused toolbar item when the user <kbd>Tab</kbd>s back to the toolbar after it is unmounted.

![Tabbing in and out the block toolbar with focus persistence](https://user-images.githubusercontent.com/3068563/95642653-3d91f780-0a80-11eb-8d1d-f4d83b486216.gif)

## How to test

1. Create any block
2. Press <kbd>Shift+Tab</kbd> to move focus to the block toolbar
3. Press arrow keys to move focus to a different toolbar button
4. Press <kbd>Tab</kbd> to move focus to the block contents
5. Do some action so as to hide the toolbar (eg. typing on the Paragraph block)
6. Press <kbd>Shift+Tab</kbd> to move focus back to the block toolbar
7. Check if the focus is persisted

Also, see the added e2e test.

## What's not covered by this PR

- The alternative discussed in https://github.com/WordPress/gutenberg/issues/24469#issuecomment-700931142 to make the toolbar only visually hidden isn't implemented in this PR. This would require a lot of changes as the app as a whole seems to rely on this mounting behavior.
- <kbd>Alt+F10</kbd> behavior is unchanged.
